File: conversions/array-flatten/elide-array-flatten.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-array-flatten

This demonstrates calling the TypeScript flatten implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One flatten library shared across TypeScript and Python
- Consistent array flattening across services
- No need for custom Python flatten implementations
- Guaranteed consistent depth handling
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# flatten = require('./elide-array-flatten.ts')

print("=== Python Calling TypeScript Array Flatten ===\n")

# Example 1: Flatten nested lists
# nested = [1, [2, [3, [4, 5]]]]
# flat = flatten.default(nested)
# print(flat)  # [1, 2, 3, 4, 5]

# Example 2: Controlled depth flattening
# def normalize_data(data: list, depth: int = 1) -> list:
#     """Flatten nested data structures"""
#     return flatten.default(data, depth)

# Example 3: Data Pipeline
# def process_batch_results(results: list[list]) -> list:
#     """Flatten batch processing results"""
#     return flatten.default(results, 1)

print("Real-world use case:")
print("- Python data pipeline processes nested batch results")
print("- Uses same TypeScript implementation as Node.js API")
print("- Guarantees consistent flattening behavior")
print("- No custom Python flatten logic needed")
print()

print("Example: Data Processing Pipeline")
print("┌─────────────────────────────────────────────┐")
print("│   Elide Array Flatten (TypeScript)        │")
print("│   elide-array-flatten.ts                   │")
print("└─────────────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │Pipeline│")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same flatten behavior everywhere!")
print()

print("Problem Solved:")
print("Before: Different flatten implementations = inconsistent data shapes")
print("After: One Elide implementation = consistent flattening")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  flatten = require('./elide-array-flatten.ts')")
print("  result = flatten.default([1, [2, [3]]])  # That's it!")


# ============================================================
# Extended Usage Examples
# ============================================================

# Example: Data Processing Pipeline
# class DataProcessor:
#     def __init__(self):
#         self.module = require('./elide-array-flatten.ts')
#     
#     def process_batch(self, data):
#         """Process data using Elide module"""
#         return self.module.default(data)
#     
#     def process_stream(self, stream):
#         """Process streaming data"""
#         results = []
#         for batch in stream:
#             results.append(self.process_batch(batch))
#         return results

# Example: API Integration
# from flask import Flask, request, jsonify
# 
# app = Flask(__name__)
# module = require('./elide-array-flatten.ts')
# 
# @app.route('/api/process', methods=['POST'])
# def process_data():
#     data = request.json.get('data')
#     result = module.default(data)
#     return jsonify({'result': result})

# Example: Async Processing
# import asyncio
# 
# async def process_async(data):
#     """Async wrapper for Elide module"""
#     loop = asyncio.get_event_loop()
#     return await loop.run_in_executor(None, module.default, data)
# 
# async def process_multiple(data_list):
#     """Process multiple items concurrently"""
#     tasks = [process_async(data) for data in data_list]
#     return await asyncio.gather(*tasks)

# Example: Caching Layer
# from functools import lru_cache
# 
# @lru_cache(maxsize=1000)
# def process_cached(data_tuple):
#     """Cached processing for frequently used inputs"""
#     return module.default(list(data_tuple))

# Example: Error Handling
# def safe_process(data):
#     """Process with comprehensive error handling"""
#     try:
#         if not data:
#             raise ValueError("Empty data provided")
#         result = module.default(data)
#         return {'success': True, 'result': result}
#     except TypeError as e:
#         return {'success': False, 'error': f'Type error: {e}'}
#     except Exception as e:
#         return {'success': False, 'error': f'Unexpected error: {e}'}

# Example: Logging and Monitoring
# import logging
# import time
# 
# logger = logging.getLogger(__name__)
# 
# def process_with_logging(data):
#     """Process with performance logging"""
#     start = time.time()
#     logger.info(f"Processing data of size: {len(data)}")
#     
#     result = module.default(data)
#     
#     duration = time.time() - start
#     logger.info(f"Processed in {duration:.4f}s")
#     return result

# Example: Testing Utilities
# import unittest
# 
# class TestModule(unittest.TestCase):
#     def setUp(self):
#         self.module = require('./elide-array-flatten.ts')
#     
#     def test_basic_functionality(self):
#         """Test basic module functionality"""
#         result = self.module.default(test_data)
#         self.assertIsNotNone(result)
#     
#     def test_edge_cases(self):
#         """Test edge case handling"""
#         self.assertRaises(TypeError, self.module.default, None)

# Example: Configuration Management
# class ConfigurableProcessor:
#     def __init__(self, config):
#         self.config = config
#         self.module = require('./elide-array-flatten.ts')
#     
#     def process(self, data):
#         """Process data according to configuration"""
#         if self.config.get('validate'):
#             self.validate_data(data)
#         
#         result = self.module.default(data)
#         
#         if self.config.get('transform'):
#             result = self.transform_result(result)
#         
#         return result
#     
#     def validate_data(self, data):
#         """Validate input data"""
#         pass
#     
#     def transform_result(self, result):
#         """Transform output"""
#         return result

print("\n" + "="*60)
print("Extended Examples Available Above")
print("="*60)
print("\nThese examples demonstrate:")
print("  • Batch processing")
print("  • API integration (Flask)")
print("  • Async/await patterns")
print("  • Caching strategies")
print("  • Error handling")
print("  • Logging and monitoring")
print("  • Unit testing")
print("  • Configuration management")
print("\nTotal: 8 comprehensive patterns")

File: conversions/array-flatten/elide-array-flatten.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-array-flatten
#
# This demonstrates calling the TypeScript flatten implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One flatten library shared across TypeScript and Ruby
# - Consistent array flattening across microservices
# - No need for custom Ruby flatten logic
# - Perfect for Rails data processing, batch operations

puts "=== Ruby Calling TypeScript Array Flatten ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# flatten_module = Elide.require('./elide-array-flatten.ts')

# Example 1: Rails Model Data Processing
# class BatchProcessor
#   def process_nested_results(results)
#     flatten_module.default(results, 1)
#   end
# end

# Example 2: Data Transformation
# class DataTransformer
#   def normalize_nested_data(data, depth = Float::INFINITY)
#     flatten_module.default(data, depth)
#   end
# end

puts "Real-world use case:"
puts "- Ruby Rails app processes nested batch results"
puts "- Uses same TypeScript implementation as Node.js API"
puts "- Guarantees consistent data shape"
puts "- No custom flatten gem needed"
puts ""

puts "Example: Batch Processing System"
puts "┌─────────────────────────────────────────────┐"
puts "│   Elide Array Flatten (TypeScript)        │"
puts "│   elide-array-flatten.ts                   │"
puts "└─────────────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │  │ Rails  │"
puts "    │  API   │  │ Worker │  │  App   │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same flattening logic everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Different flatten implementations = data shape inconsistencies"
puts "After: One Elide implementation = unified data processing"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/flatten.rb"
puts "  FLATTEN = Elide.require('./elide-array-flatten.ts')"
puts "  "
puts "  # app/services/batch_processor.rb"
puts "  def flatten_results(results)"
puts "    FLATTEN.default(results)"
puts "  end"


# ============================================================
# Extended Usage Examples
# ============================================================

# Example: Rails Service Object
# class ProcessingService
#   def initialize
#     @module = Elide.require('./elide-array-flatten.ts')
#   end
#
#   def process(data)
#     @module.default(data)
#   end
#
#   def process_batch(items)
#     items.map { |item| process(item) }
#   end
# end

# Example: Rails Controller Integration
# class DataController < ApplicationController
#   before_action :load_module
#
#   def process
#     data = params.require(:data)
#     result = @module.default(data)
#     render json: { result: result }
#   rescue => e
#     render json: { error: e.message }, status: :unprocessable_entity
#   end
#
#   private
#
#   def load_module
#     @module = Elide.require('./elide-array-flatten.ts')
#   end
# end

# Example: Background Job
# class DataProcessorJob < ApplicationJob
#   queue_as :default
#
#   def perform(data_id)
#     data = Data.find(data_id)
#     module = Elide.require('./elide-array-flatten.ts')
#     result = module.default(data.content)
#     data.update!(processed: result)
#   end
# end

# Example: ActiveRecord Extension
# module Processable
#   extend ActiveSupport::Concern
#
#   included do
#     after_save :process_data, if: :should_process?
#   end
#
#   def process_data
#     module = Elide.require('./elide-array-flatten.ts')
#     self.processed_value = module.default(self.raw_value)
#     save!
#   end
#
#   def should_process?
#     raw_value_changed?
#   end
# end

# Example: Caching with Redis
# class CachedProcessor
#   def initialize
#     @module = Elide.require('./elide-array-flatten.ts')
#     @redis = Redis.new
#   end
#
#   def process(data)
#     cache_key = Digest::SHA256.hexdigest(data.to_s)
#     cached = @redis.get(cache_key)
#     return JSON.parse(cached) if cached
#
#     result = @module.default(data)
#     @redis.setex(cache_key, 3600, result.to_json)
#     result
#   end
# end

# Example: Error Handling Wrapper
# class SafeProcessor
#   def initialize
#     @module = Elide.require('./elide-array-flatten.ts')
#     @logger = Logger.new(STDOUT)
#   end
#
#   def process(data)
#     @logger.info "Processing data"
#     result = @module.default(data)
#     @logger.info "Success"
#     { success: true, result: result }
#   rescue StandardError => e
#     @logger.error "Error: #{e.message}"
#     { success: false, error: e.message }
#   end
# end

# Example: Testing with RSpec
# RSpec.describe "Module Integration" do
#   let(:module) { Elide.require('./elide-array-flatten.ts') }
#
#   describe '#default' do
#     it 'processes data correctly' do
#       result = module.default(test_data)
#       expect(result).to be_present
#     end
#
#     it 'handles edge cases' do
#       expect { module.default(nil) }.to raise_error(TypeError)
#     end
#   end
# end

puts "\n" + "="*60
puts "Extended Examples Available Above"
puts "="*60
puts "\nThese examples demonstrate:"
puts "  • Rails service objects"
puts "  • Controller integration"
puts "  • Background jobs (ActiveJob)"
puts "  • ActiveRecord extensions"
puts "  • Redis caching"
puts "  • Error handling wrappers"
puts "  • RSpec testing"
puts "\nTotal: 7 comprehensive Rails patterns"
